tools/epoch-sync/src/cli.rs

```diff
@@ -0,0 +1,293 @@
+
+        let mut cur_hash = header_head_hash;
+
+        // Edge case if we exactly at the epoch boundary.
+        if epoch_ids.contains(&cur_hash) {
+            cur_hash = hash_to_prev_hash[&cur_hash];
+        }
+
+        let mut num_errors = 0;
+
+        while cur_hash != genesis_hash {
+            tracing::debug!("Big loop hash {:?}", cur_hash);
+
+            // epoch ids are the last hashes of some epochs
+            if epoch_ids.contains(&cur_hash) {
+                let last_header = store
+                    .get_ser::<BlockHeader>(DBCol::BlockHeader, cur_hash.as_ref())?
+                    .context("BlockHeader for cur_hash not found")?;
+                let last_finalized_height =
+                    if *last_header.last_final_block() == CryptoHash::default() {
+                        0
+                    } else {
+                        let last_finalized_header = store
+                            .get_ser::<BlockHeader>(
+                                DBCol::BlockHeader,
+                                last_header.last_final_block().as_ref(),
+                            )?
+                            .context("BlockHeader for cur_hash.last_final_block not found")?;
+                        last_finalized_header.height()
+                    };
+
+                loop {
+                    let prev_hash = hash_to_prev_hash[&cur_hash];
+                    if epoch_ids.contains(&prev_hash) {
+                        // prev_hash is the end of previous epoch
+                        // cur_hash is the start of current epoch
+                        break;
+                    } else {
+                        // prev_hash is still in the current epoch
+                        // we descent to it
+                        cur_hash = prev_hash;
+                    }
+                }
+
+                let first_block_hash = cur_hash;
+
+                let mut last_block_info = BlockInfo::new(
+                    *last_header.hash(),
+                    last_header.height(),
+                    last_finalized_height,
+                    *last_header.last_final_block(),
+                    *last_header.prev_hash(),
+                    last_header.prev_validator_proposals().collect(),
+                    last_header.chunk_mask().to_vec(),
+                    vec![],
+                    last_header.total_supply(),
+                    last_header.latest_protocol_version(),
+                    last_header.raw_timestamp(),
+                );
+
+                *last_block_info.epoch_id_mut() = last_header.epoch_id().clone();
+                *last_block_info.epoch_first_block_mut() = first_block_hash;
+
+                let next_epoch_first_hash = hash_to_next_hash[last_header.hash()];
+                tracing::debug!("Creating EpochSyncInfo from block {:?}", last_header);
+
+                let epoch_sync_info = chain_update.create_epoch_sync_info(
```

Comment:
I'm lost.
The function is called `validate`, but you're not reading any EpochSyncInfo from storage, but create them and validate the ones created?
A few comments explaining what you're validating will be helpful.

Reply:
Yeah, I will add the comment. I validate the epoch sync implementation, not the DB.